File: 1192 - Critical Connections in a Network/solution.py
class Solution:
    def criticalConnections(self, n: int, A: List[List[int]]) -> List[List[int]]:
        def dfs(idx, p, d):
            low[idx] = d
            for i in graph[idx]:
                if i == p:
                    continue
                if low[i] == -1:
                    dfs(i, idx, d+1)
                low[idx] = min(low[idx], low[i])
                if low[i] >= d+1:
                    res.append([idx, i])
        graph, low, res = [[] for _ in range(n)], [-1]*n, []
        for i, j in A:
            graph[i].append(j)
            graph[j].append(i)
        dfs(0, -1, 0)
        return res

File: 1192 - Critical Connections in a Network/solution.cpp
class Solution {
private:
    int dfs(int idx, int depth, vector<int>& rank, vector<vector<int>>& graph, vector<vector<int>>& res){
        rank[idx] = depth;
        int min_seen = depth;
        for(auto i: graph[idx]){
            if(rank[i] >= depth-1) continue;
            int seen = rank[i]==-2?dfs(i, depth+1, rank, graph, res):rank[i];
            if(seen > depth) res.push_back({idx, i});
            min_seen = min(min_seen, seen);
        }
        return min_seen;
    }
public:
    vector<vector<int>> criticalConnections(int n, vector<vector<int>>& A) {
        vector<vector<int>> graph(n), res;
        vector<int> rank(n, -2);
        for(auto& i : A){
            graph[i[0]].push_back(i[1]);
            graph[i[1]].push_back(i[0]);
        }
        dfs(0, 0, rank, graph, res);
        return res;
    }
};